dof/dof.1
\# exec: groff dof.1 -Tascii -man | less
\#
\# .TH cmd-name section [date [version [page-descr]]]
.TH dof 1 "12 Mar 2019" "v1.0" "User Commands"
.SH NAME
dof \- (do-for) executes commands for each item of a list
.SH SYNOPSIS
\# .SY command; .OP \-efp...; .OP \-d cs; .OP \-f fam; ...; .RI [ parameter .\|.\|. ]; .YS;
.SY dof
.OP \-efp
.OP \-l
.OP \-h
.OP \-v
.RI [item .\|.\|. | - ]\ do\ [ command | - ]
.SH DESCRIPTION
\fIdof\fR executes commands for each item (file or string) of a list
.PP
Example, convert all mp3s to ogg
.EX
	dof -e *.mp3 do ffmpeg -i %f %b.ogg
.EE
.SH OPTIONS
.TP
.BR \-
read from stdin.
.TP
.BR \-e
execute; \fIdof\fR displays what commands would be run, this option executes them.
.TP
.BR \-f
force non-stop; \fIdof\fR stops on error (exit code != 0), this option forces \fIdof\fR to ignore them.
.TP
.BR \-p
plain files only; directories, devices, etc are ignored.
.TP
.BR \-d
directories only; plain-files, devices, etc are ignored.
.TP
.BR \-l
print recipes (/etc/dof.conf, ~/.dof)
.TP
.BR \-\-\fIrecipe\fR
execute recipe (ex: dof --to-ogg)
.TP
.BR \-h
help screen.
.TP
.BR \-v
version and license information.
.SH VARIABLES
In the `commands' you can use several program's variables. Variables are recognized by prefix '%'.
.TP
.BR %f
the full string (or filename)
.TP
.BR %b
the basename (no directory, no extension)
.TP
.BR %d
the dirname (only the directory without the '/' suffix) 
.TP
.BR %e
the extension without '.' prefix
.TP
.BR %%
the character '%'.
.TP
.BR %(expr)
string processing expression... not used yet.
.SH FILES
dof reads \fI/etc/dof.conf\fR and \fI~/.dof\fR files.
These files contains recipes in form `name: parameters'.
These recipes can later executed with \fB--\fIname\fR option.
.PP
Example:
.EX
	to-ogg: *.mp3 do ffmpeg -i %f %b.ogg
	to-png: *.{jpg,gif} do convert %f %b.png
.EE
.SH NOTES
.TP
Items '.' and '..' are ignored.
\# .SH SEE ALSO
\# command1(section), command2(section)
.SH AUTHOR
.MT nereus@\:freemail.gr
Nicholas Christopoulos
.ME
.br
.UR https://github.com/nereusx/unix-utils
Prject page:
.UE
\# EOF


dof/dof.c
/*
 * dof [file [file [...]] do command
 * if file = - then read from stdin
 * if command = - then read from stdin
 * -e = execute, default dof run in test mode (dryrun)
 * -f = force (no stop on error)
 * -p = regular files only
 * -d = directories only
 * %f = file
 * %b = basename
 * %d = dirname
 * %e = extension
 *
 * dof 1 2 3 do echo %f
 * dof *.txt do cp %f ${dir}/%b.%e \; chown root:root ${dir}/%b.%e
 *
 * Nicholas Christopoulos (nereus@freemail.gr)
 *
 * this software is released under GPLv3 or newer license
 */

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <glob.h>

typedef struct s_node node_t;
struct s_node {
	char *str;
	char *data;
	node_t *next;
	};

typedef struct s_list list_t;
struct s_list {
	node_t *root;
	node_t *tail;
	};

static list_t file_list;
static list_t cmds_list;
static list_t rcpt_list;

#ifdef STRDUP
// clone a string
static char *strdup(const char *src)
{
	char *buf = malloc(strlen(src) + 1);
    strcpy(buf, src);
	return buf;
}
#endif

// add node to list
// this just stores a string in the list
static void list_add(list_t *list, const char *str)
{
	node_t *np = (node_t *) malloc(sizeof(node_t));
	np->str  = strdup(str);
	np->data = NULL;
	np->next = NULL;
	if ( list->root ) {
		list->tail->next = np;
		list->tail = np;
		}
	else
		list->root = list->tail = np;
}

// add node to list
// this stores a key and a value to the list
static void list_addpair(list_t *list, const char *key, const char *value)
{
	node_t *np = (node_t *) malloc(sizeof(node_t));
	np->str  = strdup(key);
	np->data = strdup(value);
	np->next = NULL;
	if ( list->root ) {
		list->tail->next = np;
		list->tail = np;
		}
	else
		list->root = list->tail = np;
}

// add node to list
// this stores wildcard matches
static void list_addwc(list_t *list, const char *pattern)
{
	glob_t globbuf;

	globbuf.gl_offs = 0;
	if ( glob(pattern, GLOB_DOOFFS, NULL, &globbuf) == 0 ) {
		for ( int i = 0; globbuf.gl_pathv[i]; i ++ ) {
			const char *name = globbuf.gl_pathv[i];
			if ( ! (strcmp(name, ".") == 0 || strcmp(name, "..") == 0) )
				list_add(list, name);
			}
		globfree(&globbuf);
		}
}

// initialize list
static void list_init(list_t *list)
{
	list->root = list->tail = NULL;
}

// clean up memory and reset the list
static void list_clear(list_t *list)
{
	node_t *pre, *cur = list->root;

	while ( cur ) {
		pre = cur;
		cur = cur->next;
		free(pre->str);
		free(pre);
		}
	list->root = list->tail = NULL;
}

// convert string list to string
static char *list_to_string(list_t *list, const char *delim)
{
	char	*ret;
	int		len = 0, delim_len = 0, first = 1;
	node_t	*cur = list->root;

	if ( delim )
		delim_len = strlen(delim);
	ret = (char *) malloc(1);
	*ret = '\0';
	while ( cur ) {
		len += delim_len + strlen(cur->str);
		ret = (char *) realloc(ret, len + 1);
		if ( first )
			first = 0;
		else if ( delim )
			strcat(ret, delim);
		strcat(ret, cur->str);
		cur = cur->next;
		}
	return ret;
}

// returns the name of the file without the directory and the extension
static char *basename(const char *source)
{
	static char buf[PATH_MAX];
	char		*b;
	const char	*p;

	if ( (p = strrchr(source, '/')) != NULL )
		strcpy(buf, p + 1);
	else
		strcpy(buf, source);
	if ( (b = (char *) strrchr(buf, '.')) != NULL )
		*b = '\0';
	return buf;
}

// returns the directory of the file without the trailing '/'
static char *dirname(const char *source)
{
	static char buf[PATH_MAX];
	const char	*p;

	if ( (p = strrchr(source, '/')) != NULL ) {
		strcpy(buf, source);
		buf[p - source] = '\0';
		}
	else
		buf[0] = '\0';
	return buf;
}

// returns the extension of the file without the '.'
static char *extname(const char *source)
{
	static char buf[PATH_MAX];
	char		*b;
	const char	*p;

	if ( (p = strrchr(source, '/')) != NULL )
		strcpy(buf, p + 1);
	else
		strcpy(buf, source);
	if ( (b = (char *) strrchr(buf, '.')) != NULL )
		return b + 1;
	else
		buf[0] = '\0';
	return buf;
}

//
static char *execblock(const char *source, const char *str)
{
	static char buf[PATH_MAX];
	char		*b;
	const char	*p;

	strcpy(buf, str);
	p = source;
	b = buf;

	// TODO: parse whole string expression,
	//		use str as input, buf as output

	// parser
/*
	while ( *p == ' ' || *p == '\t' ) p ++;
	if ( isalpha(*p) ) {
		char kw[256], *k;

		// read the keyword
		k = kw;
		while ( *p ) {
			if ( !isalpha(*p) )
				break;
			*k ++ = *p ++;
			}
		*k = '\0';

		// put 'p' to point in paramters
		while ( *p == ' ' || *p == '\t' ) p ++;

		// execute it
		if ( strcmp(kw, "replace") == 0 ) {
			// get string 1 (what)
			// get string 2 (with)
			// replace all occurences of 'what' with 'with'
			}
		}
*/
	
	return buf;
}

// converts and returns the commands to run with 'system()'
// the buffer must be freed by the caller
static char *dof(const char *fmt, const char *data)
{
	const char *p, *v;
	char *d, *dest;
	int inside_sq = 0, count = 0, maxlen, level;

	// calculate a maximum length for the returned buffer
	p = fmt;
	while ( p ) {
		count ++;
		p = strchr(p + 1, '%');
		}
	maxlen = strlen(fmt) + strlen(data) * count + 1;

	// replace strings
	dest = (char *) malloc(maxlen);
	d = dest;
	p = fmt;
	while ( *p ) {

		// handle single-quoted parts
		if ( inside_sq ) {
			if ( *p == '\'' ) {
				inside_sq = 0;
				p ++;
				continue;
				}
			*d ++ = *p ++;
			continue;
			}
		if ( *p == '\'' ) {
			inside_sq = 1;
			p ++;
			continue;
			}

		// translate % expressions
		if ( *p == '%' ) {
			char block[LINE_MAX], *bp;
			
			p ++;
			switch ( *p ) {
			case 'f':		// full pathname
				v = data;
				while ( *v )	*d ++ = *v ++;
				break;
			case 'b':		// basename (no directory, no extension)
				v = basename(data);
				while ( *v )	*d ++ = *v ++;
				break;
			case 'd':		// directory name
				v = dirname(data);
				while ( *v )	*d ++ = *v ++;
				break;
			case 'e':		// extension
				v = extname(data);
				while ( *v )	*d ++ = *v ++;
				break;
			case '(':		// block
				p ++;
				bp = block;
				level = 0;
				while ( *p ) {
					if ( *p == ')' ) {
						level --;
						if ( level < 0 )
							break;
						}
					else if ( *p == '(' )
						level ++;
					*bp ++ = *p ++;
					}
				*bp = '\0';
				v = execblock(block, data);
				while ( *v )	*d ++ = *v ++;
				break;			
			case '%':		// none
				*d ++ = *p;
				break;
			default:
				fprintf(stderr, "unknown element `%%%c'\n", *p);
				}
			}
		// not a command? just copy
		else
			*d ++ = *p;

		// next
		if ( *p )
			p ++;
		}

	// close destination string
	*d = '\0';

	return dest;
}

// read configuration files
static void	read_conf()
{
	FILE	*fp;
	char	buf[LINE_MAX], *p, *key, *data;
	char	file[PATH_MAX];
	static	int readconf_init = 0;

	if ( readconf_init ) return;
	readconf_init = 1;
	
	for ( int i = 0; i < 2; i ++ ) {
		switch ( i ) {
		case 0: strcpy(file, "/etc/dof.conf"); break;
		case 1: strcpy(file, getenv("HOME")); strcat(file, "/.dof"); break;
			}
		if ( (fp = fopen(file, "rt")) != NULL ) {
			while ( fgets(buf, LINE_MAX, fp) ) {
				p = buf;
				while ( *p == ' ' || *p == '\t' ) p ++;
				if ( *p == '\n' || *p == '\0' || *p == '#' )
					continue;
				key = p;
				while ( *p ) {
					if ( *p == ':' )
						break;
					p ++;
					}
				if ( *p == ':' ) {
					*p = '\0';
					p ++;
					while ( *p == ' ' || *p == '\t' ) p ++;
					data = p;
					if ( data[strlen(data)-1] == '\n' )
						data[strlen(data)-1] = '\0';
					list_addpair(&rcpt_list, key, data);
					}
				}
			fclose(fp);
			}
		}
}

// --- main() ---

#define APP_DESCR \
"dof (do-for) run commands for each element of 'list'."

static const char *usage = "\
Usage: dof [list] do [commands]\n\
"APP_DESCR"\n\
\n\
Options:\n\
\t-e\texecute; dof displays what commands would be run, this option executes them.\n\
\t-f\tforce non-stop; dof stops on error, this option forces dof to ignore errors.\n\
\t-p\tplain files only; directories, devices, etc are ignored.\n\
\t-d\tdirectories only; plain files, devices, etc are ignored.\n\
\t-l\tprint recipes (/etc/dof.conf, ~/.dof)\n\
\t--<recipe>\n\t\texecute recipe (ex: dof --to-ogg)\n\
\t-\tread from stdin\n\
\t-h\tthis screen\n\
\t-v\tversion and program information\n\
\n\
Variables:\n\
\t%f\tthe string (if file; the full path name)\n\
\t%b\tthe basename (no directory, no extension)\n\
\t%d\tthe directory (without trailing '/')\n\
\t%e\tthe extension (without '.')\n\
";

static const char *verss = "\
dof version 1.00\n\
"APP_DESCR"\n\
\n\
Copyright (C) 2017 Free Software Foundation, Inc.\n\
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.\n\
This is free software: you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.\n\
\n\
Written by Nicholas Christopoulos <mailto:nereus@freemail.gr>\n\
";

// return a pointer to filename without the directory
static const char *namep(const char *file)
{
	const char *p;
	if ( (p = strrchr(file, '/')) != NULL )
		return p + 1;
	return file;
}

// returns true if the "filename" has wildcards
static int has_wildcards(const char *filename)
{
	const char *p = filename;
	
	while ( *p ) {
		if ( *p == '*' || *p == '?' || *p == '[' )
			return 1;
		p ++;
		}
	return 0;
}

//
int main(int argc, char **argv)
{
	int		flags = 0, state = 0, exit_status = 0;
	node_t	*cur;
	char	*cmds, *cmdbuf;
	struct stat st;

	list_init(&file_list);
	list_init(&cmds_list);
	list_init(&rcpt_list);
	for ( int i = 1; i < argc; i ++ ) {
		if ( argv[i][0] == '-' && state == 0 ) {

			// one minus, read from stdin
			if ( argv[i][1] == '\0' ) {
				char buf[LINE_MAX];

				while ( fgets(buf, LINE_MAX, stdin) )	{
					if ( state == 0 ) {
						if (
							strcmp(namep(argv[i]), ".") == 0 ||
							strcmp(namep(argv[i]), "..") == 0
							)
							; // ignore them
						else
							list_add(&file_list, buf);
						}
					else
						list_add(&cmds_list, buf);
					}
				}

			// check options
			for ( int j = 1; argv[i][j]; j ++ ) {
				switch ( argv[i][j] ) {
				case 'e': flags |= 0x01; break;
				case 'f': flags |= 0x02; break;
				case 'p': flags |= 0x04; break;
				case 'd': flags |= 0x08; break;
				case 'h': puts(usage); return 1;
				case 'v': puts(verss); return 1;
				case '-':
						// execute recipe
						read_conf();
						cur = rcpt_list.root;
						while ( cur ) {
							if ( strcmp(cur->str, argv[i]+2) == 0 ) {
								char cmd[LINE_MAX];
								char opt[64];
								opt[0] = '\0';
								if ( flags & 0x01 ) strcat(opt, "-e ");
								if ( flags & 0x02 ) strcat(opt, "-f ");
								if ( flags & 0x04 ) strcat(opt, "-p ");
								if ( flags & 0x08 ) strcat(opt, "-d ");
								snprintf(cmd, LINE_MAX, "dof %s %s", opt, cur->data);
								return system(cmd);
								}
							cur = cur->next;
							}
						return 1;
				case 'l':
						// list recipes
						read_conf();
						cur = rcpt_list.root;
						while ( cur ) {
							printf("%s: (%s)\n", cur->str, cur->data);
							cur = cur->next;
							}
						return 0;
				default:
					// error
					puts("unknown option.");
					return 1;
					}
				}
			}
		else if ( strcmp(argv[i], "do") == 0 && state == 0 )
			state ++;
		else {
			if ( state == 0 ) {
				if (
					strcmp(namep(argv[i]), ".") == 0 ||
					strcmp(namep(argv[i]), "..") == 0
					)
					;
				else {
					if ( has_wildcards(argv[i]) )
						list_addwc(&file_list, argv[i]);
					else
						list_add(&file_list, argv[i]);
					}
				}
			else
				list_add(&cmds_list, argv[i]);
			}
		}

	if ( state == 0 ) { // syntax error
		puts("`do' keyword missing; run `dof -h' for help.");
		return 1;
		}

	// execute
	cmds = list_to_string(&cmds_list, " ");
	cur = file_list.root;
	while ( cur ) {
		if ( flags & 0x04 || flags & 0x08 ) { // file attribute check
			if ( stat(cur->str, &st) == 0 ) {
				if ( flags & 0x04 ) { // plain files only
					if ( ! S_ISREG(st.st_mode) ) {
						cur = cur->next;
						continue;
						}
					}
				if ( flags & 0x08 ) { // directories only
					if ( ! S_ISDIR(st.st_mode) ) {
						cur = cur->next;
						continue;
						}
					}
				}
			}
		cmdbuf = dof(cmds, cur->str);
		if ( (flags & 0x01) == 0 ) // not execute-option
			fprintf(stdout, "%s\n", cmdbuf);
		else {
			if ( (exit_status = system(cmdbuf)) != 0 ) {
				if ( (flags & 0x02) == 0 ) { // not force-option
					free(cmdbuf);
					break;
					}
				}
			}
		free(cmdbuf);
		cur = cur->next;
		}

	// cleanup
	list_clear(&file_list);
	list_clear(&cmds_list);
	list_clear(&rcpt_list);
	return exit_status;
}
